### DISCUSSION TITLE
Why does setting the log level programmatically not work? `with_max_level` not working?

### DISCUSSION THREAD

**Discussion by bluenote10** [tokio-rs/tracing#2551] [Q&A] [answered]
I'm trying for a while to set the log level programmatically. The very first lines of my `main` function are:

```rs
tracing_subscriber::fmt::fmt()
    .with_max_level(tracing::Level::DEBUG)
    .compact()
    .finish();

tracing::info!("Test info level");
tracing::debug!("Test debug level");
```

I would have expected to see these traces on stdout, but there is nothing. Is this a bug or am I misunderstanding something here?

**Comment by hawkw** [comment, marked as answer]
You aren't [setting the `Subscriber` as the default](https://docs.rs/tracing/latest/tracing/dispatcher/index.html). The [`fmt::SubscriberBuilder::finish` ](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.finish) method you're calling _returns_ a new `fmt::Subscriber`, but it doesn't actually configure `tracing` to use that `Subscriber` to collect traces. Try passing the returned `fmt::Subscriber` to [`tracing::subscriber::set_global_default`](https://docs.rs/tracing/latest/tracing/subscriber/fn.set_global_default.html), or call [`SubscriberBuilder::init()`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.init) rather than `SubscriberBuilder::finish` (which calls `set_global_default` for you), and then your code should work the way you expect.

We should probably add `#[must_use]` annotations to either the `finish` method or the `fmt::Subscriber` type, so that this code would have generated a warning telling you you're not using the returned subscriber...then, the problem would have hopefully been a lot more obvious.
